admin/client.py

```diff
@@ -1,176 +1,164 @@
-    """
-    Interface for starting and stopping nodes for acceptance testing.
-    """
-
-    def start_nodes(reactor):
+    def __init__(self, effects):
+        self.lines = [
+            '#!/bin/bash',
+            'set -ex'
+        ]
+        TypeDispatcher.__init__(self, {
+            Run: self.perform_run,
+            Sudo: perform_sudo,
+            Put: perform_put,
+            Comment: self.perform_comment,
+            Sequence: perform_sequence
+        })
+        perform(self, effects)
+        # Add blank line to terminate script with a newline
+        self.lines.append('')
+        self._script = '\n'.join(self.lines)
+
+    @sync_performer
+    def perform_run(self, dispatcher, intent):
         """
-        Start nodes for running acceptance tests.
-
-        :param reactor: Reactor to use.
-        :return Deferred: Deferred which fires with a list of nodes to run
-            tests against.
+        For Run effects, add the command line.
         """
+        self.lines.append(intent.command)
 
-    def stop_nodes(reactor):
+    @sync_performer
+    def perform_comment(self, dispatcher, intent):
         """
-        Stop the nodes started by `start_nodes`.
-
-        :param reactor: Reactor to use.
-        :return Deferred: Deferred which fires when the nodes have been
-            stopped.
+        For Comment effects, prefix the comment with #
         """
+        self.lines.append('# ' + intent.comment)
 
+    def script(self):
+        """
+        Return the generated shell script.
+        """
+        return self._script
 
-RUNNER_ATTRIBUTES = [
-    'distribution', 'top_level', 'config', 'package_source'
-]
 
+def make_script_file(dir, effects):
+    """
+    Create a shell script file from a set of effects.
 
-@implementer(INodeRunner)
-@attributes(RUNNER_ATTRIBUTES + [
-    'provisioner',
-], apply_immutable=True)
-class LibcloudRunner(object):
+    :param dir: The directory in which to create the script.
+    :param effects: The effects which contain the commands.
+    :return: The base filename of the script.
     """
-    Start and stop cloud nodes for acceptance testing.
+    builder = ScriptBuilder(effects)
+    fd, filename = tempfile.mkstemp(dir=dir, text=True)
+    os.write(fd, builder.script())
+    os.close(fd)
+    os.chmod(filename, 0555)
+    return os.path.basename(filename)
+
 
-    :ivar LibcloudProvioner provisioner: The provisioner to use to create the
-        nodes.
-    :ivar VolumeBackend volume_backend: The volume backend the nodes are
-        configured with.
+class DockerRunner:
+    """
+    Run commands on Docker.
     """
-    def __init__(self):
-        self.nodes = []
 
-        self.metadata = self.config.get('metadata', {})
-        try:
-            creator = self.metadata['creator']
-        except KeyError:
-            raise UsageError("Must specify creator metadata.")
+    def __init__(self, image):
+        self.docker = docker.Client(version='1.18')
```

Comment:
Our Ubuntu build slaves use Docker 1.3, which only supports 1.15. Since these tests run on a CentOS buildslave, this was not detected. Found while converting to Jenkins.